src/modules/hip/kernel/down_mixing.hpp

```diff
@@ -0,0 +1,73 @@
+#include <hip/hip_runtime.h>
+#include "rpp_hip_common.hpp"
+
+__global__ void down_mixing_hip_tensor(float *srcPtr,
+                                       uint srcStride,
+                                       float *dstPtr,
+                                       uint dstStride,
+                                       int2 *srcDimsTensor)
+
+{
+    int id_x = hipBlockIdx_x * hipBlockDim_x + hipThreadIdx_x;
+    int id_z = hipBlockIdx_z * hipBlockDim_z + hipThreadIdx_z;
+    int srcLength = srcDimsTensor[id_z].x;
+    int channels = srcDimsTensor[id_z].y;
+
+    if (id_x >= srcLength)
+        return;
+
+    float nomalizedWeight = 1.f / channels;
+    float outVal = 0.0f;
+    uint srcIdx = id_z * srcStride + id_x * channels;
+    int i = 0;
+    int alignedChannels = (channels / 8) * 8;
+    // if number of channels is a multiple of 8, do 8 pixel load till alignedChannels value
+    if (alignedChannels)
+    {
+        d_float8 outVal_f8;
+        outVal_f8.f4[0] = static_cast<float4>(0.0f);
+        outVal_f8.f4[1] = outVal_f8.f4[0];
+        float4 normalizedWeight_f4 = static_cast<float4>(nomalizedWeight);
+        for(; i < alignedChannels; i += 8, srcIdx += 8)
+        {
+            d_float8 src_f8;
+            rpp_hip_load8_and_unpack_to_float8(srcPtr + srcIdx, &src_f8);
+            rpp_hip_math_multiply8_const(&src_f8, &src_f8, normalizedWeight_f4);
+            rpp_hip_math_add8(&outVal_f8, &src_f8, &outVal_f8);
+        }
+        outVal_f8.f4[0] += outVal_f8.f4[1];
+        outVal += (outVal_f8.f1[0] + outVal_f8.f1[1] + outVal_f8.f1[2] + outVal_f8.f1[3]);
+    }
+    // process remaining channels
+    for(; i < channels; i++, srcIdx++)
+        outVal += srcPtr[srcIdx] * nomalizedWeight;
+
+    uint dstIdx = id_z * dstStride + id_x;
+    dstPtr[dstIdx] = outVal;
```

Comment:
What is the range of values channels we can have? Any ways to vectorize the write here?

Reply:
I see that the RPPT_MAX_AUDIO_CHANNELS = 16 in the Resample PR. If thats the case, possibly running the 1 liner un-vectorized loop on L43 to do a "outVal += srcPtr[srcIdx]" should be fine followed by doing a "outVal *= nomalizedWeight" at the end.
What could possibly be tried is if you:
- reduce threads launched by a factor of 8 (the usual >> 3 with d_float8s, or try with a >> 2 launch with float4s)
- vectorize the dst writes here so we write 8 outVal_f8
- to compute outVal_f8, do a "outVal_f8 += srcVals_f8" in the loop
- resetting srcVals_f8 each time (channels times) inside the loop from global mem could take time, but I suggest trying this

Reply:
Tried this experiment and it is giving a good 30% improvement in performance

Previously
I did not vectorize the writes because, for the higher channel cases (16 channel input), we need to load 128 elements per thread (8 * 16) and felt it increases the load too much per thread

Currently we have only 2 channel input in test suite and it goes to non vectorized loop, so it is loading 16 elements per thread and giving a 30% boost compared to unvectorized writes case

Reply:
I think this performance improvement was an anomoly
Reran the tests for updating numbers in MS, but observing 3x more time for vectorized writes version compared to unvectorized writes

Reran multiple times today and confirmed there was a degradation in performance
So reverted back to single pixel writes version